### PR TITLE
[SYCL][HIP] Add basic HIP atomics

### PR DESCRIPTION
**Pull request by hdelan**
Adding support for basic atomic operations for HIP AMD backend.

### CODE REVIEW

**Comment by hdelan** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1510

File: clang/lib/Frontend/InitPreprocessor.cpp [hunk 1]
```diff
@@ -588,17 +588,19 @@ static void InitializeStandardPredefinedMacros(const TargetInfo &TI,
...
-    Builder.defineMacro("__HIP_MEMORY_SCOPE_WORKGROUP", "3");
-    Builder.defineMacro("__HIP_MEMORY_SCOPE_AGENT", "4");
-    Builder.defineMacro("__HIP_MEMORY_SCOPE_SYSTEM", "5");
     if (LangOpts.CUDAIsDevice)
       Builder.defineMacro("__HIP_DEVICE_COMPILE__");
     if (LangOpts.GPUDefaultStream ==
         LangOptions::GPUDefaultStreamKind::PerThread)
       Builder.defineMacro("HIP_API_PER_THREAD_DEFAULT_STREAM");
   }
+  if (LangOpts.HIP || (LangOpts.OpenCL && TI.getTriple().isAMDGPU())) {
```
Comment (premanandrao) [inline clang/lib/Frontend/InitPreprocessor.cpp:597]: Please add a test for this functionality.

File: libclc/amdgcn-amdhsa/libspirv/atomic/atomic_add.cl [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+              __hip_atomic_fetch_add)
+
+AMDGPU_ATOMIC(_Z21__spirv_AtomicFAddEXT, float, f, global, AS1,
+              __hip_atomic_fetch_add)
+AMDGPU_ATOMIC(_Z21__spirv_AtomicFAddEXT, float, f, local, AS3,
+              __hip_atomic_fetch_add)
+
+// TODO implement for fp64
+
+#undef AMDGPU_ATOMIC
```
Comment (steffenlarsen) [inline libclc/amdgcn-amdhsa/libspirv/atomic/atomic_add.cl:38]: Should `GET_ATOMIC_SCOPE_AND_ORDER` also be undefined here?

File: sycl/include/sycl/atomic_ref.hpp [hunk 3]
```diff
@@ -126,6 +126,12 @@ class atomic_ref_base {
       detail::IsValidDefaultOrder<DefaultOrder>::value,
       "Invalid default memory_order for atomics.  Valid defaults are: "
       "relaxed, acq_rel, seq_cst");
+#ifdef __AMDGPU__
+  // FIXME should this query device's memory capabilities at runtime?
+  static_assert("seq_cst memory order is not supported on AMDGPU" &&
+                DefaultOrder != sycl::memory_order::seq_cst);
```
Comment (steffenlarsen) [inline sycl/include/sycl/atomic_ref.hpp:132]:
```suggestion
  static_assert(DefaultOrder != sycl::memory_order::seq_cst,
                "seq_cst memory order is not supported on AMDGPU");
```

File: clang/lib/Frontend/InitPreprocessor.cpp (hunk 1 above)
Comment (elizabethandrews) [reply clang/lib/Frontend/InitPreprocessor.cpp:597]: And for `SYCL_USE_NATIVE_FP_ATOMICS`

File: libclc/amdgcn-amdhsa/libspirv/atomic/atomic_add.cl (hunk 2 above)
Comment (hdelan) [reply libclc/amdgcn-amdhsa/libspirv/atomic/atomic_add.cl:21]: Libclc files updated, thanks

File: sycl/include/sycl/atomic_ref.hpp (hunk 3 above)
Comment (al42and) [reply sycl/include/sycl/atomic_ref.hpp:132]: I think the idea is to use `static_assert(cond, msg)`, not `static_assert(cond && msg)`.

**Review by steffenlarsen — APPROVED** [review]
Good stuff! 🚀

File: clang/lib/Frontend/InitPreprocessor.cpp (hunk 1 above)
Comment (hdelan) [reply clang/lib/Frontend/InitPreprocessor.cpp:597]: Tests added

File: clang/test/Preprocessor/opencl-macro-target-specific.cl [hunk 4]
```diff
@@ -0,0 +1,7 @@
+// RUN: %clang_cc1 %s -E -dM -triple amdgcn-amdhsa-amdhsa \
+// RUN: | FileCheck --check-prefix=CHECK-AMDGPU %s
```
Comment (premanandrao) [inline clang/test/Preprocessor/opencl-macro-target-specific.cl:2]: Please add a comment here about what the test does.

File: clang/test/Preprocessor/sycl-macro-target-specific.cpp [hunk 5]
```diff
@@ -8,9 +8,26 @@
 // RUN: | FileCheck --check-prefix=CHECK-NVPTX-NEG %s
 // RUN: %clang_cc1 %s -fsycl-is-device -triple spir64_fpga-unknown-unknown -E -dM \
 // RUN: | FileCheck --check-prefix=CHECK-NVPTX-NEG %s
+// RUN: %clang_cc1 %s -fsycl-is-device -triple amdgcn-amdhsa-amdhsa -E -dM \
+// RUN: | FileCheck --check-prefix=CHECK-NVPTX-NEG %s
 // CHECK-NVPTX: #define __NVPTX__
 // CHECK-NVPTX-NEG-NOT: #define __NVPTX__
```
Comment (premanandrao) [inline clang/test/Preprocessor/sycl-macro-target-specific.cpp:14]: Here too, though I understand this is not a new test.

**Review by premanandrao — COMMENTED** [review]
A couple of comments needed, but otherwise FE changes LGTM.

File: clang/test/Preprocessor/sycl-macro-target-specific.cpp (hunk 5 above)
Comment (hdelan) [reply clang/test/Preprocessor/sycl-macro-target-specific.cpp:16]: Comment added

File: clang/test/Preprocessor/opencl-macro-target-specific.cl (hunk 4 above)
Comment (hdelan) [reply clang/test/Preprocessor/opencl-macro-target-specific.cl:4]: Comment added

**Review by premanandrao — APPROVED** [review]

**Review by Fznamznon — APPROVED** [review]
Clang changes look ok to me.

**Comment by hdelan** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1510

**Comment by zjin-lcf** [comment]
Faster than hipcc ... 

SYCL
./main 10000
Atomic min for data type U64 | Average execution time: 0.000025 (s)
Atomic min for data type S64 | Average execution time: 0.000021 (s)
Atomic min for data type F64 | Average execution time: 0.000022 (s)
Atomic max for data type U64 | Average execution time: 0.000052 (s)
Atomic max for data type S64 | Average execution time: 0.000054 (s)
Atomic max for data type F64 | Average execution time: 0.000052 (s)
Atomic add for data type U64 | Average execution time: 3.235254 (s)
Atomic add for data type S64 | Average execution time: 3.215109 (s)
Atomic add for data type F64 | Average execution time: 3.245661 (s)
atomic max results: 262144 -9223372036854775808 262144.000000
FAIL

=================================================
HIP
./main 10000
Atomic min for data type U64 | Average execution time: 0.000053 (s)
Atomic min for data type S64 | Average execution time: 0.000029 (s)
Atomic min for data type F64 | Average execution time: 0.000029 (s)
Atomic max for data type U64 | Average execution time: 0.000280 (s)
Atomic max for data type S64 | Average execution time: 0.000292 (s)
Atomic max for data type F64 | Average execution time: 0.000280 (s)
Atomic add for data type U64 | Average execution time: 4.652767 (s)
Atomic add for data type S64 | Average execution time: 4.517521 (s)
Atomic add for data type F64 | Average execution time: 4.601057 (s)
PASS

**Comment by hdelan** [comment]
@zjin-lcf wow nice. Note that AtomicFMin and AtomicFMax are not included in this PR, which is why the tests are failing.

**Comment by hdelan** [comment]
@bader the LLVM test suite failures are all XPASSes, so I think this can be merged

**Comment by zjin-lcf** [comment]
> @zjin-lcf wow nice. Note that AtomicFMin and AtomicFMax are not included in this PR, which is why the tests are failing.

Is there something missing in ROCm for not including FMin/FMax ?

**Comment by hdelan** [comment]
No it is just lower priority than the FAdd etc. I will be adding FMin, FMax as soon as I have finished some other stuff I am working on

**Comment by bader** [comment]
> @bader the LLVM test suite failures are all XPASSes, so I think this can be merged

I see these failures:

Failed Tests (4):
  SYCL :: Reduction/reduction_big_data.cpp
  SYCL :: Reduction/reduction_nd_ext_double.cpp
  SYCL :: Reduction/reduction_usm.cpp
  SYCL :: Reduction/reduction_usm_dw.cpp

lld: error: undefined hidden symbol: __spirv_AtomicFMaxEXT(float AS1*, __spv::Scope::Flag, __spv::MemorySemanticsMask::Flag, float)
>>> referenced by lto.tmp:(typeinfo name for sycl::_V1::detail::reduction::MainKrn<main::A1, (sycl::_V1::detail::reduction::strategy)4>)
>>> referenced by lto.tmp:(typeinfo name for sycl::_V1::detail::reduction::MainKrn<main::A1, (sycl::_V1::detail::reduction::strategy)4>)
>>> referenced by lto.tmp:(_ZTSN4sycl3_V16detail9reduction7MainKrnIZ4mainE2A1LNS2_8strategyE4EJEEE_with_offset)
>>> referenced 1 more times
llvm-foreach: 
clang-16: error: amdgcn-link command failed with exit code 1 (use -v to see invocation)

error: command failed with exit status: 1

These seem to be related to the change.

NOTE: Unfortunately, "/verify with" doesn't cover AMDGPU, so there is no safe way to test both changes (compiler + test) in CI.

**Comment by hdelan** [comment]
@bader that should be fixed now. @zjin-lcf remaining atomics are all added, so FMax, FMin etc should be working

**Comment by zjin-lcf** [comment]
I tried to run atomic tests. Can you please run them again ? Thanks. 

https://github.com/zjin-lcf/HeCBench/tree/master/atomicCAS-sycl

**Comment by hdelan** [comment]
> I tried to run atomic tests. Can you please run them again ? Thanks.
> 
> https://github.com/zjin-lcf/HeCBench/tree/master/atomicCAS-sycl

```
$ sycl-ls
[ext_oneapi_hip:gpu:0] AMD HIP BACKEND, AMD Radeon VII 0.0 [HIP 40421.43]
$ ./main 10000
Atomic min for data type U64 | Average execution time: 0.000072 (s)
Atomic min for data type S64 | Average execution time: 0.000087 (s)
Atomic min for data type F64 | Average execution time: 0.000061 (s)
Atomic max for data type U64 | Average execution time: 0.000218 (s)
Atomic max for data type S64 | Average execution time: 0.000226 (s)
Atomic max for data type F64 | Average execution time: 0.000250 (s)
Atomic add for data type U64 | Average execution time: 2.484193 (s)
Atomic add for data type S64 | Average execution time: 2.413692 (s)
Atomic add for data type F64 | Average execution time: 2.370539 (s)
PASS
```

**Comment by zjin-lcf** [comment]
@hdelan 
I pushed some changes to the sycl benchmark to fix some issues, and observed the same result (PASS) on a MI200 GPU. Congratulations!

I tried to run the benchmark 
https://github.com/zjin-lcf/HeCBench/tree/master/sheath-hip
https://github.com/zjin-lcf/HeCBench/tree/master/sheath-sycl

The option `-munsafe-fp-atomics` is passed to the hip compiler and the option `-DSYCL_USE_NATIVE_FP_ATOMICS` to the sycl compiler. However, I think that the sycl compiler does not generate fast fp atomics due to the significant difference in execution time on a MI200 GPU.  Do I pass the right compiler option to the sycl compiler ?  Thanks again.

**Comment by bader** [comment]
> so I think this can be merged

One more thing. We need to merge these changes together with test changes, but the test PR is not reviewed yet.

**Comment by hdelan** [comment]
> @hdelan I pushed some changes to the sycl benchmark to fix some issues, and observed the same result (PASS) on a MI200 GPU. Congratulations!
> 
> I tried to run the benchmark https://github.com/zjin-lcf/HeCBench/tree/master/sheath-hip https://github.com/zjin-lcf/HeCBench/tree/master/sheath-sycl
> 
> The option `-munsafe-fp-atomics` is passed to the hip compiler and the option `-DSYCL_USE_NATIVE_FP_ATOMICS` to the sycl compiler. However, I think that the sycl compiler does not generate fast fp atomics due to the significant difference in execution time on a MI200 GPU. Do I pass the right compiler option to the sycl compiler ? Thanks again.

@zjin-lcf So the fast gfx90a+ atomics are not part of this PR. They will be enabled in a future PR, will keep you posted

**Comment by zjin-lcf** [comment]
I see. I am also puzzled by the issue (https://github.com/intel/llvm/issues/7252).
The atomic add works with "shared" USM, but other atomics (e.g. atomic add, atomic xor) don't.

**Comment by hdelan** [comment]
> 

Can we get the test suite changes reviewed?

**Comment by hdelan** [comment]
@bader the test suite changes have been reviewed so I think this is safe to merge

**Pull request merged**